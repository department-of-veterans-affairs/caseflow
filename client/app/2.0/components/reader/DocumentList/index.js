export * from 'components/reader/DocumentList/ClaimsFolderDetails';
export * from 'components/reader/DocumentList/CommentsTable';
export * from 'components/reader/DocumentList/DocumentsTable';
export * from 'components/reader/DocumentList/Header';
export * from 'components/reader/DocumentList/LastRetrievalAlert';
export * from 'components/reader/DocumentList/LastRetrievalInfo';
export * from 'components/reader/DocumentList/NoSearchResults';
export * from 'components/reader/DocumentList/QueueLink';
