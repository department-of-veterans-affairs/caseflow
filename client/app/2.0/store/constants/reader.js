import React from 'react';

// Local Dependencies
import { CategoryIcon } from 'app/components/RenderFunctions';

export const CATEGORIES = {
  VIEW_DOCUMENT_PAGE: 'Document Viewer',
  CLAIMS_FOLDER_PAGE: 'Claims Folder',
  CASE_SELECTION_PAGE: 'Case Selection'
};

export const ACTION_NAMES = {
  VIEW_NEXT_DOCUMENT: 'view-next-document',
  VIEW_PREVIOUS_DOCUMENT: 'view-previous-document'
};

export const INTERACTION_TYPES = {
  VISIBLE_UI: 'visible-ui',
  KEYBOARD_SHORTCUT: 'keyboard-shortcut'
};

export const ENDPOINT_NAMES = {
  DOCUMENT: 'document',
  ANNOTATION: 'annotation',
  MARK_DOC_AS_READ: 'mark-doc-as-read',
  TAG: 'tag',
  APPEAL_DETAILS: 'appeal-details',
  APPEAL_DETAILS_BY_VET_ID: 'appeal-details-by-vet-id',
  CLAIMS_FOLDER_SEARCHES: 'claims-folder-searches',
  DOCUMENTS: 'documents'
};

export const COMMENT_ACCORDION_KEY = 'Comments';

// If we used CSS in JS, we wouldn't have to keep this value in sync with the CSS in a brittle way.
export const ANNOTATION_ICON_SIDE_LENGTH = 40;

// Arrange the directions such that each direction + 1 modulo 4 rotates clockwise.
export const MOVE_ANNOTATION_ICON_DIRECTIONS = {
  ArrowUp: 0,
  ArrowRight: 1,
  ArrowDown: 2,
  ArrowLeft: 3
};

// An array with the directions oriented in a clockwise rotation.
export const MOVE_ANNOTATION_ICON_DIRECTION_ARRAY = [
  MOVE_ANNOTATION_ICON_DIRECTIONS.ArrowUp,
  MOVE_ANNOTATION_ICON_DIRECTIONS.ArrowRight,
  MOVE_ANNOTATION_ICON_DIRECTIONS.ArrowDown,
  MOVE_ANNOTATION_ICON_DIRECTIONS.ArrowLeft
];

export const documentCategories = {
  procedural: {
    renderOrder: 0,
    humanName: 'Procedural',
    svg: <CategoryIcon color="#5A94EC" />
  },
  medical: {
    renderOrder: 1,
    humanName: 'Medical',
    svg: <CategoryIcon color="#FF6868" />
  },
  other: {
    renderOrder: 2,
    humanName: 'Other Evidence',
    svg: <CategoryIcon color="#3AD2CF" />
  },
  case_summary: {
    renderOrder: 3,
    humanName: 'Case Summary',
    svg: <CategoryIcon color="#FDC231" />,
    readOnly: true
  }
};

// UI Text
export const NO_ISSUES_ON_APPEAL_MSG = 'No issues on appeal';

// These both come from _pdf_viewer.css and is the default height
// of the pages in the PDF. We need it defined here to be
// able to expand/contract the height of the pages as we zoom.
export const PDF_PAGE_WIDTH = 816;
export const PDF_PAGE_HEIGHT = 1056;

// Defined in _search_bar.scss > .cf-search-bar. Max height of
// search bar, used as offset when scrolling to search results
export const SEARCH_BAR_HEIGHT = 50;

// This comes from the class .pdfViewer.singlePageView .page in _reviewer.scss.
// We need it defined here to be able to expand/contract margin between pages
// as we zoom.
export const PAGE_MARGIN = 25;

// Base scale used to calculate dimensions and draw text.
export const PAGE_DIMENSION_SCALE = 1;

export const DOCUMENTS_OR_COMMENTS_ENUM = {
  DOCUMENTS: 'documents',
  COMMENTS: 'comments'
};

// Rotation Constants
export const ROTATION_INCREMENTS = 90;
export const COMPLETE_ROTATION = 360;

// Documents Table Constants
export const NUMBER_OF_COLUMNS = 6;

// PDF Viewer Constatns
export const NUMBER_OF_DIRECTIONS = 4;

// Zoom Constants
export const ZOOM_RATE = 0.3;
export const MINIMUM_ZOOM = 0.1;

export const LOADING_DATA_MESSAGE = 'Loading claims folder in Reader...';

// Set the cache timeout
export const CACHE_TIMEOUT_HOURS = 3;

export const COMMENT_SCROLL_FROM_THE_TOP = 50;
