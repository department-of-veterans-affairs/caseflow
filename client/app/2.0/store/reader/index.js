// External Dependencies
import { combineReducers } from 'redux';

// Reducers
import appeal from 'store/reader/appeal';
import documentList from 'store/reader/documentList';
import documentViewer from 'store/reader/documentViewer';
import annotationLayer from 'store/reader/annotationLayer';

/**
 * Root Reader Reducer
 */
const readerReducer = combineReducers({
  appeal,
  documentViewer,
  documentList,
  annotationLayer
});

/**
 * Default export wraps the reducer in an HOC for capturing metrics
 */
export default readerReducer;
