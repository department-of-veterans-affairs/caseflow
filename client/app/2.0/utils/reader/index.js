/**
 * Helper Method to update the focus of the Documents Table
 * @param {element} lastReadRef -- React ref to the current Last Read Indicator
 * @param {element} tbodyRef -- React ref to the current Table Body
 */
export const focusElement = (lastReadRef, tbodyRef) => {
  // Set the Initial Scroll position
  let scrollTop = tbodyRef.scrollTop;

  // Focus the Last Read Indicator if present
  if (lastReadRef) {
    // Get the Last Read Indicator Boundary
    const lastReadContainer = lastReadRef.getBoundingClientRect();

    // Get the Table Body Boundary
    const tbodyContainer = tbodyRef.getBoundingClientRect();

    // Check if the Last Read Indicator is in view based on whether it is in the table body boundary
    if (tbodyContainer.top >= lastReadContainer.top && lastReadContainer.bottom >= tbodyContainer.bottom) {
      // Find the row to focus
      const rowWithLastRead = find(tbodyRef.children, (tr) => tr.querySelector(`#${lastReadRef.id}`));

      // Update the scroll position to focus the Last Read Row
      scrollTop += rowWithLastRead.getBoundingClientRect().top - tbodyContainer.top;
    }
  }

  // Return the Scroll Position to update the table
  return scrollTop;
};

export * from 'utils/reader/format';
export * from 'utils/reader/search';
export * from 'utils/reader/pdf';
export * from 'utils/reader/document';
export * from 'utils/reader/comments';
export * from 'utils/reader/coordinates';
export * from 'utils/reader/keyboard';
export * from 'utils/reader/user';
export * from 'utils/reader/appeal';
