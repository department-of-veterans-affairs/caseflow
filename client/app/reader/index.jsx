import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import DecisionReviewer from './DecisionReviewer';
import readerReducer from './reducer';

export const initialState = {
  ui: {
    pdf: {},
    pdfSidebar: {
      showTagErrorMsg: false,
      hidePdfSidebar: false
    },
    pdfList: {
      lastReadDocId: null,
      filters: {
        category: {}
      },
      dropdowns: {
        category: false
      }
    }
  },
  documents: {}
};

// eslint-disable-next-line no-underscore-dangle
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store =
  createStore(readerReducer, initialState, composeEnhancers(applyMiddleware(thunk)));

if (module.hot) {
  // Enable Webpack hot module replacement for reducers
  module.hot.accept('./reducer', () => {
    store.replaceReducer(readerReducer);
  });
}

const Reader = (props) => {
  return <Provider store={store}>
      <DecisionReviewer {...props} />
  </Provider>;
};

export default Reader;
