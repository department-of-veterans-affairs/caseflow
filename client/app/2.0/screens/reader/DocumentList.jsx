// External Dependencies
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';

// Local Dependencies
import { recordSearch, fetchDocuments } from 'utils/reader';
import { documentListScreen } from 'store/reader/selectors';
import {
  setSearch,
  clearSearch as clear,
  changeView,
  changeSortState,
  setCategoryFilter,
  toggleDropdownFilterVisibility,
  clearCategoryFilters,
  clearAllFilters,
  clearTagFilters,
  setTagFilter,
  toggleComment,
} from 'store/reader/documentList';

import {
  QueueLink,
  ClaimsFolderDetails,
  DocumentListHeader,
  LastRetrievalAlert,
  LastRetrievalInfo,
  NoSearchResults,
  DocumentsTable,
  CommentsTable,
} from 'components/reader/DocumentList';
import { selectComment } from 'store/reader/annotationLayer';

const DocumentList = (props) => {
  // Get the Document List state
  const state = useSelector(documentListScreen);

  // Create the Dispatcher
  const dispatch = useDispatch();

  // Load the Documents
  useEffect(fetchDocuments({ ...state, params: props.match.params }, dispatch), []);

  // Create the dispatchers
  const actions = {
    toggleComment: (docId, expanded) => dispatch(toggleComment({ docId, expanded })),
    clearTagFilters: () => dispatch(clearTagFilters(state)),
    setTagFilter: (text, checked) => dispatch(setTagFilter(text, checked, state)),
    clearAllFilters: () => dispatch(clearAllFilters(state)),
    clearCategoryFilters: () => dispatch(clearCategoryFilters(state)),
    toggleFilter: (val) => dispatch(toggleDropdownFilterVisibility(val)),
    setCategoryFilter: (categoryName, checked) => dispatch(setCategoryFilter(categoryName, checked, state)),
    changeView: (val) => dispatch(changeView(val)),
    search: (val) => dispatch(setSearch(val, state.comments, state.storeDocuments)),
    clearSearch: () => dispatch(clear(state.filterCriteria, state.comments, state.storeDocuments)),
    recordSearch: (query) => recordSearch(props.match.params.vacolsId, query),
    changeSort: (val) => dispatch(changeSortState(val, state)),
    selectComment: (comment) => {
      // Scroll to the comment location
      // focusComment(comment);

      // Update the store with the selected component
      dispatch(selectComment(comment));
    }
  };

  return (
    <React.Fragment>
      {state.queueRedirectUrl && (
        <QueueLink {...state} veteranFullName={state.appeal.veteran_full_name} vbmsId={state.appeal.vbms_id} />
      )}
      <AppSegment filledBackground>
        <div className="section--document-list">
          <ClaimsFolderDetails {...state} />
          <LastRetrievalAlert {...state.documentList} appeal={state.appeal} />
          <DocumentListHeader {...state} {...actions} />
          <NoSearchResults {...state} {...actions} show={state.documentsView === 'none'} />
          <DocumentsTable
            {...state}
            {...actions}
            documentPathBase={`/reader/appeal/${ state.appeal.id }/documents`}
            show={state.documentsView === 'documents'}
          />
          <CommentsTable
            {...state}
            {...actions}
            documentPathBase={`/reader/appeal/${ state.appeal.id }/documents`}
            show={state.documentsView === 'comments'}
          />
        </div>
      </AppSegment>
      <LastRetrievalInfo {...state.documentList} />
    </React.Fragment>
  );
};

DocumentList.propTypes = {
  appeal: PropTypes.object,
  pdfWorker: PropTypes.string,
  userDisplayName: PropTypes.string,
  dropdownUrls: PropTypes.array,
  singleDocumentMode: PropTypes.bool,
  match: PropTypes.object,
  annotations: PropTypes.array,

  // Required actions
  setCategoryFilter: PropTypes.func,
};

export default DocumentList;
