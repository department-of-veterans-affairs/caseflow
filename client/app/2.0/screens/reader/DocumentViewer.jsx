// External Dependencies
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import classNames from 'classnames';

// Local Dependencies
import { getPageCoordinatesOfMouseEvent } from 'utils/reader';
import { pdfWrapper } from 'styles/reader/Document/PDF';
import { fetchDocuments, openDownloadLink } from 'utils/reader/document';
import { documentScreen } from 'store/reader/selectors';
import { DocumentHeader } from 'components/reader/DocumentViewer/Header';
import { DocumentSidebar } from 'components/reader/DocumentViewer/Sidebar';
import { DocumentFooter } from 'components/reader/DocumentViewer/Footer';
import { DocumentSearch } from 'app/2.0/components/reader/DocumentViewer/Search';
import { Pdf } from 'app/2.0/components/reader/DocumentViewer/PDF';
import { ZOOM_RATE, MINIMUM_ZOOM, CATEGORIES, PAGE_MARGIN } from 'app/2.0/store/constants/reader';
import { ShareComment } from 'app/2.0/components/reader/DocumentViewer/modals/Share';
import { DeleteComment } from 'app/2.0/components/reader/DocumentViewer/modals/Delete';
import {
  showPdf,
  togglePdfSideBar,
  toggleSearchBar,
  toggleAccordion,
  toggleShareModal,
  toggleDeleteModal,
  setOverscanValue,
  saveDescription,
  changeDescription,
  resetDescription,
  handleCategoryToggle,
  setPageNumber,
  searchText,
  toggleKeyboardInfo,
  addTag,
  removeTag,
  clearSearch,
} from 'store/reader/documentViewer';
import {
  selectComment,
  startEdit,
  updateComment,
  saveComment,
  addComment,
  cancelDrop,
  dropComment,
  createComment,
  removeComment,
  startMove,
  moveComment
} from 'store/reader/annotationLayer';
import { KeyboardInfo } from 'app/2.0/components/reader/DocumentViewer/modals/KeyboardInfo';

/**
 * Document Viewer Screen Component
 * @param {Object} props -- Contains the route props and PDF web worker
 */
const DocumentViewer = (props) => {
  // Get the Document List state
  const state = useSelector(documentScreen);

  // Create the Dispatcher
  const dispatch = useDispatch();

  // Attach the PDF Worker to the params to setup PDFJS
  const params = {
    ...props.match.params,
    worker: props.pdfWorker,
    currentDocument: state.currentDocument,
    scale: state.scale
  };

  // Calculate the Next and Prev Docs
  const docs = {
    prev: state.filteredDocIds[state.filteredDocIds.indexOf(state.currentDocument.id) - 1],
    current: state.filteredDocIds.indexOf(state.currentDocument.id),
    next: state.filteredDocIds[state.filteredDocIds.indexOf(state.currentDocument.id) + 1],
  };

  // Create the Grid Ref
  const gridRef = React.createRef();

  // Create the dispatchers
  const actions = {
    changeTags: (values, deleted) => {
      // Delete tags if there are any removed
      if (deleted) {
        // Pull the value to delete out of the list
        const [tag] = deleted;

        // Request the removal of the selected tags
        dispatch(removeTag({
          doc: state.currentDocument,
          tag: state.currentDocument.tags.reduce((list, item) => item.text === tag.label ? item : list, {})
        }));
      } else if (values?.length) {
        // Set the tags to create
        const tags = values.
          filter((value) => !state.currentDocument.tags.map((tag) => tag.text).includes(value.label)).
          map((tag) => ({ text: tag.label }));

        // Request the creation of the new tags
        dispatch(addTag({
          doc: state.currentDocument,
          tags
        }));
      }
    },
    getCoords: (event, pageIndex) => getPageCoordinatesOfMouseEvent(
      event,
      document.getElementById(`comment-layer-${pageIndex}`).getBoundingClientRect(),
      state.scale,
      state.currentDocument.rotation
    ),
    dropComment: (coords, pageIndex) => {
      // Drop the comment at the coordinates
      if (state.addingComment) {
        dispatch(dropComment({
          document_id: state.currentDocument.id,
          pendingComment: '',
          id: 'placing-annotation-icon',
          page: pageIndex + 1,
          x: coords.x,
          y: coords.y,
        }));
      }
    },
    moveComment: (coords, pageIndex) => {
      // Assign the current icon component
      const icon = document.getElementById(`commentIcon-container-${state.movingComment}`);

      // Calculate the x and y offset by the icon height and width
      const x = (coords.x * state.scale);
      const y = (coords.y * state.scale);

      // Move the Comment in the UI immediately
      icon.style.left = `${x}px`;
      icon.style.top = `${y}px`;

      // Request the move, if failed the comment will revert otherwise it will catch up
      dispatch(moveComment({
        page: pageIndex + 1,
        document_id: state.currentDocument.id,
        id: state.movingComment,
        x,
        y,
      }));
    },
    moveMouse: (coords, pageIndex) => {
      if (state.addingComment) {
        // Move the cursor icon
        const cursor = document.getElementById(`canvas-cursor-${pageIndex}`);

        // Hide all the other cursors
        Array.from(document.getElementsByClassName('canvas-cursor')).forEach((canvas) => {
          canvas.style.display = 'none';
        });

        // Update the coordinates
        cursor.style.left = `${coords.x * state.scale}px`;
        cursor.style.top = `${coords.y * state.scale}px`;
        cursor.style.display = 'block';
      }
    },
    showPdf: (currentPage, currentDocument, scale) => dispatch(showPdf({
      currentDocument,
      pageNumber: currentPage,
      worker: props.pdfWorker,
      scale
    })),
    toggleKeyboardInfo: (val) => dispatch(toggleKeyboardInfo(val)),
    startMove: (commentId) => dispatch(startMove(commentId)),
    clickPage: (event) => {
      event.stopPropagation();
      event.preventDefault();

      if (state.addingComment) {
        dispatch(cancelDrop());
      }

      if (state.selectedComment) {
        dispatch(selectComment({}));
      }
    },
    cancelDrop: () => dispatch(cancelDrop()),
    addComment: () => dispatch(addComment()),
    saveComment: (comment, action = 'save') => {
      // Construct a whitespace-only Regex
      const whitespace = new RegExp(/^\s+$/);

      // Handle empty comments
      if ((whitespace.test(comment.pendingComment) || whitespace.test(comment.comment)) && action === 'save') {
        return dispatch(toggleDeleteModal(comment.id));
      }

      // Calculate the comment data to update/create
      const data = {
        ...comment,
        relevant_date: comment.pendingDate || comment.relevant_date,
        comment: comment.pendingComment || comment.comment
      };

      // Send the analytics event based on the action
      window.analyticsEvent(CATEGORIES.VIEW_DOCUMENT_PAGE, `request-${action}-annotation`);

      // Determine whether to save or create the comment
      const dispatcher = action === 'create' ? createComment : saveComment;

      // Update the store to either save or create the comment
      dispatch(dispatcher(data));
    },
    updateComment: (comment) => dispatch(updateComment(comment)),
    editComment: (commentId) => dispatch(startEdit(commentId)),
    resetEdit: () => {
      dispatch(updateComment({}));
      dispatch(startEdit(null));
    },
    selectComment: (comment) => dispatch(selectComment(comment)),
    deselectComment: () => {
      if (state.selectedComment) {
        dispatch(selectComment({}));
      }
    },
    searchText: (searchTerm, index) => {
      // Calculate the match index
      const maxIndex = index >= state.search.totalMatchesInFile ? 0 : index;
      const matchIndex = index < 0 ? state.search.totalMatchesInFile - 1 : maxIndex;

      dispatch(searchText({ searchTerm, docId: state.currentDocument.id, matchIndex }));
    },
    saveDescription: (description) => dispatch(saveDescription({ docId: state.currentDocument.id, description })),
    changeDescription: (description) => dispatch(changeDescription(description)),
    resetDescription: () => dispatch(resetDescription()),
    setOverscanValue: (val) => dispatch(setOverscanValue(val)),
    handleCategoryToggle: (categoryKey, toggleState) => dispatch(handleCategoryToggle({
      docId: state.currentDocument.id,
      categoryKey,
      toggleState
    })),
    closeShareModal: () => dispatch(toggleShareModal(null)),
    closeDeleteModal: () => dispatch(toggleDeleteModal(null)),
    shareComment: (id) => dispatch(toggleShareModal(id)),
    deleteComment: (id) => dispatch(toggleDeleteModal(id)),
    removeComment: () => {
      // Send the analytics event
      window.analyticsEvent(CATEGORIES.VIEW_DOCUMENT_PAGE, 'request-delete-annotation');

      dispatch(removeComment({ commentId: state.deleteCommentId, docId: state.currentDocument.id }));
    },
    toggleAccordion: (sections) => dispatch(toggleAccordion(sections, state.openSections)),
    togglePdfSidebar: (open = null) => dispatch(togglePdfSideBar(open)),
    toggleSearchBar: (open = null) => {
      // Toggle the Search
      dispatch(toggleSearchBar(open));

      // Clear the term
      dispatch(searchText({ searchTerm: '', docId: state.currentDocument.id, matchIndex: 0 }));
    },
    download: () => openDownloadLink(state.currentDocument.content_url, state.currentDocument.type),
    scrollPage: ({ scrollTop }) => {
      // Calculate the Page Offset
      const offset = Math.floor(scrollTop / state.viewport.height);

      // Set the Current page number
      const pageNumber = offset + 1;

      // Calculate the current page
      const currentPage = pageNumber > gridRef.current?.props?.rowCount ? gridRef.current?.props?.rowCount : pageNumber;

      // Update the Pages if the client height and canvas list have changed
      if (pageNumber !== state.currentDocument.currentPage) {
        dispatch(setPageNumber(currentPage));
      }

    },
    overscanIndices: ({ cellCount, overscanCellsCount, startIndex, stopIndex }) => ({
      overscanStartIndex: Math.max(0, startIndex - Math.ceil(overscanCellsCount / 2)),
      overscanStopIndex: Math.min(cellCount - 1, stopIndex + Math.ceil(overscanCellsCount / 2))
    }),
    fitToScreen: () => {
      window.analyticsEvent(CATEGORIES.VIEW_DOCUMENT_PAGE, 'fit to screen');

      dispatch(showPdf({
        currentDocument: state.currentDocument,
        worker: props.pdfWorker,
        scale: 1
      }));
    },
    rotateDocument: () => {
      dispatch(showPdf({
        currentDocument: state.currentDocument,
        rotation: state.currentDocument.rotation,
        worker: props.pdfWorker,
        scale: state.scale
      }));
    },
    zoom: (direction) => {
      // Determine the Amount to zoom based on the direction
      const delta = direction === 'in' ? ZOOM_RATE : -ZOOM_RATE;

      // Calculate the new Scale to zoom the document
      const scale = Math.min(Math.max(MINIMUM_ZOOM, state.scale + delta), 2);

      window.analyticsEvent(CATEGORIES.VIEW_DOCUMENT_PAGE, `zoom ${direction}`, scale);

      dispatch(showPdf({ currentDocument: state.currentDocument, worker: props.pdfWorker, scale }));
    },
    setPageNumber: (pageNumber) => {
      // Add the analytics event
      window.analyticsEvent(CATEGORIES.VIEW_DOCUMENT_PAGE, 'jump-to-page');

      // Calculate the page number
      const number = pageNumber - 1;
      const page = number >= gridRef.current?.props?.rowCount ? gridRef.current?.props?.rowCount - 1 : number;

      // Scroll to the page
      gridRef.current?.scrollToPosition({ scrollTop: page * (state.viewport.height + PAGE_MARGIN) });
    },
    prevDoc: () => {
      // Clear the document search
      dispatch(clearSearch());

      const doc = state.documents[docs.prev];

      // Load the previous doc if found
      if (doc) {
        props.history.push(`/reader/appeal/${params.vacolsId}/documents/${doc.id}`);
      }
    },
    nextDoc: () => {
      // Clear the document search
      dispatch(clearSearch());

      const doc = state.documents[docs.next];

      // Load the next doc if found
      if (doc) {
        props.history.push(`/reader/appeal/${params.vacolsId}/documents/${doc.id}`);
      }
    }
  };

  // Load the Documents
  useEffect(() => {
    // Get the Current document
    const currentDocument = state.documents[params.docId];

    // Load the PDF
    if (currentDocument?.id) {
      dispatch(showPdf({
        currentDocument,
        worker: props.pdfWorker,
        scale: state.scale
      }));
    } else {
      // Load the Documents
      fetchDocuments({ ...state, params }, dispatch)();
    }
  }, [params.docId]);

  return (
    <div id="document-viewer" className="cf-pdf-page-container" onClick={actions.deselectComment} >
      <div className={classNames('cf-pdf-container', { 'hidden-sidebar': state.hidePdfSidebar })} {...pdfWrapper}>
        <DocumentHeader {...props} {...state} {...actions} doc={state.currentDocument} />
        <DocumentSearch {...actions} {...state.search} doc={state.currentDocument} hidden={state.hideSearchBar} />
        <Pdf {...props} {...state}{...actions} doc={state.currentDocument} gridRef={gridRef} />
        <DocumentFooter
          {...props}
          {...state}
          {...actions}
          nextDocId={docs.next > 0 ? docs.next : 0}
          prevDocId={docs.prev > 0 ? docs.prev : 0}
          currentIndex={docs.current}
          doc={state.currentDocument}
        />
      </div>
      <DocumentSidebar {...props} {...state} {...actions} show={!state.hidePdfSidebar} doc={state.currentDocument} />
      <ShareComment {...state} {...actions} show={state.shareCommentId !== null} commentId={state.shareCommentId} />
      <DeleteComment {...state} {...actions} show={state.deleteCommentId !== null} />
      <KeyboardInfo {...state} {...actions} show={state.keyboardInfoOpen} />
    </div>
  );
};

DocumentViewer.propTypes = {
  appeal: PropTypes.object,
  history: PropTypes.object,
  pdfWorker: PropTypes.string,
  userDisplayName: PropTypes.string,
  dropdownUrls: PropTypes.array,
  singleDocumentMode: PropTypes.bool,
  match: PropTypes.object,
  annotations: PropTypes.array,
};

export default DocumentViewer;
