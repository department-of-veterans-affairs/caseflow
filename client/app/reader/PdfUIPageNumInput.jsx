import React from 'react';
import PropTypes from 'prop-types';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { jumpToPage } from './actions';
import { isValidWholeNumber } from './utils';
import TextField from '../components/TextField';

const ENTER_KEY = 'Enter';
const RADIX = 10;

export class PdfUIPageNumInput extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      pageNumber: 1
    };
  }

  componentWillUpdate = (nextProps) => {
    if (nextProps.currentPage !== this.props.currentPage) {
      this.setPageNumber(nextProps.currentPage);
    }
  }

  setPageNumber = (pageNumber) => {
    this.setState({
      pageNumber
    });
  }

  handleKeyPress = (event) => {
    if (event.key === ENTER_KEY) {
      const pageNumber = event.target.value;
      const newPageNumber = this.validatePageNum(pageNumber);

      this.setPageNumber(newPageNumber);
      // don't jump to the page unless it's a valid page entry
      // and it's not the current page
      if (this.props.currentPage !== newPageNumber) {
        this.props.jumpToPage(newPageNumber, this.props.docId);
      }
    }
  }

  validatePageNum = (pageNumber) => {
    let pageNum = parseInt(pageNumber, RADIX);

    if (!pageNum || !isValidWholeNumber(pageNum) ||
      (pageNum < 1 || pageNum > this.props.numPages)) {
      return this.props.currentPage;
    }

    return pageNum;
  }

  render() {
    return (
    <div style={{ display: 'inline-block' }}>
      <TextField
        maxLength="4"
        name="page-progress-indicator-input"
        label="Page"
        onChange={this.setPageNumber}
        onKeyPress={this.handleKeyPress}
        value={this.state.pageNumber}
        required={false}
        className={['page-progress-indicator-input']}
      />
    </div>);
  }
}

PdfUIPageNumInput.propTypes = {
  currentPage: PropTypes.number,
  numPages: PropTypes.number,
  jumpToPage: PropTypes.func,
  docId: PropTypes.number
};

const mapDispatchToProps = (dispatch) => bindActionCreators({
  jumpToPage
}, dispatch);

export default connect(
  null, mapDispatchToProps
)(PdfUIPageNumInput);
