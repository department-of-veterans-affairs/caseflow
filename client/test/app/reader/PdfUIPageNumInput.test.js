import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent, { specialChars } from '@testing-library/user-event';
import { PdfUIPageNumInput } from '../../../app/reader/PdfUIPageNumInput';

describe('PdfUIPageNumInput', () => {
  const jumpToPage = jest.fn();

  const defaultProps = {
    jumpToPage,
    numPages: 20,
    docId: 42
  };

  const setupTestComponent = (props = {}) => {
    return render(<PdfUIPageNumInput {...defaultProps} {...props} />);
  };

  beforeEach(() => {
    jumpToPage.mockClear();
  });

  it('renders properly', () => {
    const component = setupTestComponent();

    expect(component).toMatchSnapshot();
  });

  it('shows default page number of 1', () => {
    setupTestComponent();

    expect(screen.getByLabelText('Page').value).toEqual('1');
  });

  describe('jump to page', () => {
    it('jumps to new page when different page number input', () => {
      setupTestComponent();
      const input = screen.getByLabelText('Page');

      userEvent.type(input, `${specialChars.backspace}3${specialChars.enter}`);

      expect(jumpToPage).toHaveBeenCalledTimes(1);
      expect(screen.getByLabelText('Page').value).toEqual('3');
    });

    it('parses non int input to int and jumps to that page', () => {
      setupTestComponent();
      const input = screen.getByLabelText('Page');

      userEvent.type(input, `${specialChars.backspace}2.5${specialChars.enter}`);

      expect(jumpToPage).toHaveBeenCalledTimes(1);
      expect(screen.getByLabelText('Page').value).toEqual('2');
    });

    it('does not jump to new page if input is greater than number of pages', () => {
      setupTestComponent();
      const input = screen.getByLabelText('Page');

      userEvent.type(input, `${specialChars.backspace}100${specialChars.enter}`);

      expect(jumpToPage).toHaveBeenCalledTimes(0);
    });

    it('does not jump to new page if input is not a number', () => {
      setupTestComponent();
      const input = screen.getByLabelText('Page');

      userEvent.type(input, `${specialChars.backspace}foo${specialChars.enter}`);

      expect(jumpToPage).toHaveBeenCalledTimes(0);
    });
  });
});
