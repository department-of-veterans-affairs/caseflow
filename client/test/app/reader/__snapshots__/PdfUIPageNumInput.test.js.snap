// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PdfUIPageNumInput renders properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: inline-block;"
      >
        <div
          class="page-progress-indicator-input   "
        >
          <label
            for="page-progress-indicator-input"
          >
            <span>
              Page
            </span>
          </label>
          <div
            class="input-container"
          >
            <input
              class="page-progress-indicator-input"
              id="page-progress-indicator-input"
              maxlength="4"
              name="page-progress-indicator-input"
              type="text"
              value="1"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: inline-block;"
    >
      <div
        class="page-progress-indicator-input   "
      >
        <label
          for="page-progress-indicator-input"
        >
          <span>
            Page
          </span>
        </label>
        <div
          class="input-container"
        >
          <input
            class="page-progress-indicator-input"
            id="page-progress-indicator-input"
            maxlength="4"
            name="page-progress-indicator-input"
            type="text"
            value="1"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
